File: sdks/sdkwork-models-sdk/sdkwork-models-sdk-typescript/src/loaders.ts
import type { ModelCatalog, VendorCatalog } from "./types.js";

export async function loadCatalog(root: string): Promise<ModelCatalog> {
  const source = catalogSource(root);
  const manifest = await readJson(source, "sdkwork-models.json");
  const meters = (await readJson(source, "models/meters.json")).meters;
  const protocols = (await readJson(source, "models/protocols.json")).protocols;
  const index = await readJson(source, "models/index.json");
  const vendors: VendorCatalog[] = [];
  for (const vendor of index.vendors ?? []) {
    vendors.push(await loadVendorCatalog(root, vendor.vendorCode, vendor.regionCode));
  }
  return {
    catalogVersion: manifest.catalogVersion,
    schemaVersion: manifest.schemaVersion,
    meters,
    protocols,
    vendors,
  };
}

export async function loadVendorCatalog(root: string, vendorCode: string, regionCode: string): Promise<VendorCatalog> {
  const source = catalogSource(root);
  const index = await readJson(source, "models/index.json");
  const vendorIndex = index.vendors.find(
    (item: { vendorCode: string; regionCode: string }) => item.vendorCode === vendorCode && item.regionCode === regionCode,
  );
  if (!vendorIndex) {
    throw new Error(`vendor region ${vendorCode}/${regionCode} is not indexed`);
  }
  const vendor = await readJson(source, `models/${vendorIndex.path}`);
  const models = await Promise.all((vendorIndex.modelFiles ?? []).map((path: string) => readJson(source, `models/${path}`)));
  const pricing = await Promise.all((vendorIndex.pricingFiles ?? []).map((path: string) => readJson(source, `models/${path}`)));
  return { vendorCode, regionCode, vendor, models, pricing };
}

export async function loadBundledCatalog(): Promise<ModelCatalog> {
  const configuredRoot = runtimeEnv().SDKWORK_MODELS_CATALOG_ROOT;
  if (configuredRoot && configuredRoot.trim().length > 0) {
    return loadCatalog(configuredRoot);
  }
  return loadCatalog(await resolveDefaultCatalogRoot());
}

interface CatalogSource {
  root: string;
  remote: boolean;
}

function catalogSource(root: string): CatalogSource {
  return { root: root.replace(/[\\/]+$/, ""), remote: isRemoteUrl(root) };
}

function isRemoteUrl(value: string): boolean {
  try {
    const url = new URL(value);
    return url.protocol === "http:" || url.protocol === "https:";
  } catch {
    return false;
  }
}

function runtimeEnv(): Record<string, string | undefined> {
  return (globalThis as { process?: { env?: Record<string, string | undefined> } }).process?.env ?? {};
}

async function resolveDefaultCatalogRoot(): Promise<string> {
  const fs = await import("node:fs");
  const { dirname, join } = await import("node:path");
  const { fileURLToPath } = await import("node:url");
  let current = dirname(fileURLToPath(import.meta.url));
  for (let depth = 0; depth < 8; depth += 1) {
    if (fs.existsSync(join(current, "sdkwork-models.json"))) {
      return current;
    }
    const parent = dirname(current);
    if (parent === current) {
      break;
    }
    current = parent;
  }
  return "data/sdkwork-models";
}

async function readJson(source: CatalogSource, path: string): Promise<any> {
  if (source.remote) {
    const response = await fetch(`${source.root}/${path}`);
    if (!response.ok) {
      throw new Error(`failed to fetch sdkwork-models catalog file ${path}: ${response.status}`);
    }
    return response.json();
  }
  const { readFile } = await import("node:fs/promises");
  const { join } = await import("node:path");
  return JSON.parse(await readFile(join(source.root, ...path.split("/")), "utf8"));
}

File: sdks/sdkwork-models-sdk/sdkwork-models-sdk-python/sdkwork_models/loaders.py
from __future__ import annotations

import json
import os
from pathlib import Path
from urllib.parse import urljoin, urlparse
from urllib.request import urlopen

from .types import ModelCatalog


def _is_remote_url(value: str | Path) -> bool:
    parsed = urlparse(str(value))
    return parsed.scheme in {"http", "https"}


def _read_json(root: str | Path, rel_path: str) -> dict:
    if _is_remote_url(root):
        with urlopen(urljoin(f"{str(root).rstrip('/')}/", rel_path), timeout=30) as response:
            return json.loads(response.read().decode("utf-8"))
    return json.loads((Path(root) / rel_path).read_text(encoding="utf-8"))


def load_catalog(path_or_url: str | Path) -> ModelCatalog:
    manifest = _read_json(path_or_url, "sdkwork-models.json")
    meters = _read_json(path_or_url, "models/meters.json")["meters"]
    protocols = _read_json(path_or_url, "models/protocols.json")["protocols"]
    index = _read_json(path_or_url, "models/index.json")
    vendors: list[dict] = []
    vendor_codes: set[str] = set()
    vendor_catalogs: list[dict] = []
    models: dict[str, dict] = {}
    model_scores: dict[str, int] = {}
    pricing: list[dict] = []
    for item in index.get("vendors", []):
        vendor_code = item["vendorCode"]
        region_code = item["regionCode"]
        vendor_catalog = load_vendor_catalog(path_or_url, vendor_code, region_code)
        vendor_catalogs.append(vendor_catalog)
        if vendor_code not in vendor_codes:
            vendor_codes.add(vendor_code)
            vendors.append(vendor_catalog["vendor"])
        for model in vendor_catalog["models"]:
            _put_best_model_identity(models, model_scores, vendor_catalog, model)
        pricing.extend(vendor_catalog["pricing"])
    return ModelCatalog(
        catalog_version=manifest["catalogVersion"],
        schema_version=manifest["schemaVersion"],
        meters=meters,
        protocols=protocols,
        vendors=vendors,
        vendor_catalogs=vendor_catalogs,
        models=list(models.values()),
        pricing=pricing,
    )


def load_bundled_catalog() -> ModelCatalog:
    configured_root = os.getenv("SDKWORK_MODELS_CATALOG_ROOT", "").strip()
    if configured_root:
        return load_catalog(configured_root)
    for ancestor in Path(__file__).resolve().parents:
        if (ancestor / "sdkwork-models.json").is_file():
            return load_catalog(ancestor)
    return load_catalog(Path("data") / "sdkwork-models")


def load_vendor_catalog(path_or_url: str | Path, vendor_code: str, region_code: str) -> dict:
    index = _read_json(path_or_url, "models/index.json")
    vendor_index = next(
        (
            item
            for item in index.get("vendors", [])
            if item.get("vendorCode") == vendor_code and item.get("regionCode") == region_code
        ),
        None,
    )
    if vendor_index is None:
        raise ValueError(f"vendor region {vendor_code}/{region_code} is not indexed")
    voices_path = vendor_index.get("voicesPath")
    voices: list[dict] = []
    if isinstance(voices_path, str) and voices_path:
        voices = _read_json(path_or_url, f"models/{voices_path}").get("voices", [])
    model_voice_bindings = [
        _read_json(path_or_url, f"models/{path}") for path in vendor_index.get("modelVoiceFiles", [])
    ]
    model_video_profiles = [
        _read_json(path_or_url, f"models/{path}") for path in vendor_index.get("modelVideoProfileFiles", [])
    ]
    return {
        "vendorCode": vendor_code,
        "regionCode": region_code,
        "vendor": _read_json(path_or_url, f"models/{vendor_index['path']}"),
        "families": _read_json(path_or_url, f"models/{vendor_index['familiesPath']}"),
        "models": [_read_json(path_or_url, f"models/{path}") for path in vendor_index.get("modelFiles", [])],
        "pricing": [_read_json(path_or_url, f"models/{path}") for path in vendor_index.get("pricingFiles", [])],
        "voices": voices,
        "modelVoiceBindings": model_voice_bindings,
        "modelVideoProfiles": model_video_profiles,
    }


def _put_best_model_identity(models: dict[str, dict], scores: dict[str, int], vendor_catalog: dict, model: dict) -> None:
    catalog_key = model.get("catalogKey")
    if not isinstance(catalog_key, str) or not catalog_key:
        return
    score = _model_identity_score(vendor_catalog, model)
    if catalog_key not in models or score > scores[catalog_key]:
        models[catalog_key] = model
        scores[catalog_key] = score


def _model_identity_score(vendor_catalog: dict, model: dict) -> int:
    score = 0
    if _has_region_pricing(vendor_catalog, model):
        score += 100
    if model.get("routingState") == "enabled":
        score += 40
    if model.get("shelfState") == "listed":
        score += 20
    if model.get("releaseStage") == "active":
        score += 10
    if model.get("lifecycle") in {"current", "preview"}:
        score += 5
    if model.get("regionCode") == "global":
        score += 1
    return score


def _has_region_pricing(vendor_catalog: dict, model: dict) -> bool:
    model_id = model.get("modelId")
    return isinstance(model_id, str) and any(
        item.get("modelId") == model_id and item.get("prices")
        for item in vendor_catalog.get("pricing", [])
    )
